File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        result = list()
        result.append(nums[0])

        for i in range(1, len(nums)):
            if result[-1] < nums[i]:
                result.append(nums[i])
            else:
                lo, hi = 0, len(result) - 1

                while lo < hi:
                    mid = (lo + hi) // 2

                    if result[mid] < nums[i]:
                        lo = mid + 1
                    else:
                        hi = mid

                result[lo] = nums[i]

        return len(result)

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> result = new ArrayList<>();

        for (int num: nums) {
            if (result.isEmpty()) {
                result.add(num);
                continue;
            }
            if (num > result.get(result.size() - 1)) {
                result.add(num);
                continue;
            }
            int lo = 0;
            int hi = result.size() - 1;

            while (lo < hi) {
                int mid = (lo + hi) / 2;
                if (result.get(mid) < num) {
                    lo = mid + 1;
                } else {
                    hi = mid;
                }
            }
            result.set(lo, num);
        }

        return result.size();
    }
}